pkg/ingester/circuitbreaker.go

```diff
@@ -0,0 +1,249 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ingester
+
+import (
+	"context"
+	"flag"
+	"time"
+
+	"github.com/failsafe-go/failsafe-go/circuitbreaker"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/grpcutil"
+	"github.com/grafana/dskit/middleware"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"google.golang.org/grpc/codes"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+)
+
+type testCtxKey string
+
+const (
+	resultSuccess            = "success"
+	resultError              = "error"
+	resultOpen               = "circuit_breaker_open"
+	testDelayKey  testCtxKey = "test-delay"
+
+	circuitBreakerCurrentStateGaugeName  = "cortex_ingester_circuit_breaker_current_state"
+	circuitBreakerCurrentStateGaugeHelp  = "Boolean set to 1 whenever the circuit breaker is in a state corresponding to the label name."
+	circuitBreakerCurrentStateGaugeLabel = "state"
+)
+
+type circuitBreakerMetrics struct {
+	circuitBreakerOpenStateGauge     prometheus.GaugeFunc
+	circuitBreakerHalfOpenStateGauge prometheus.GaugeFunc
+	circuitBreakerClosedStateGauge   prometheus.GaugeFunc
+
+	circuitBreakerTransitions *prometheus.CounterVec
+	circuitBreakerResults     *prometheus.CounterVec
+}
+
+func newCircuitBreakerMetrics(r prometheus.Registerer, currentStateFn func() circuitbreaker.State) *circuitBreakerMetrics {
+	return &circuitBreakerMetrics{
+		circuitBreakerOpenStateGauge: promauto.With(r).NewGaugeFunc(prometheus.GaugeOpts{
+			Name:        circuitBreakerCurrentStateGaugeName,
+			Help:        circuitBreakerCurrentStateGaugeHelp,
+			ConstLabels: map[string]string{circuitBreakerCurrentStateGaugeLabel: circuitbreaker.OpenState.String()},
+		}, func() float64 {
+			if currentStateFn() == circuitbreaker.OpenState {
+				return 1
+			}
+			return 0
+		}),
+		circuitBreakerHalfOpenStateGauge: promauto.With(r).NewGaugeFunc(prometheus.GaugeOpts{
+			Name:        circuitBreakerCurrentStateGaugeName,
+			Help:        circuitBreakerCurrentStateGaugeHelp,
+			ConstLabels: map[string]string{circuitBreakerCurrentStateGaugeLabel: circuitbreaker.HalfOpenState.String()},
+		}, func() float64 {
+			if currentStateFn() == circuitbreaker.HalfOpenState {
+				return 1
+			}
+			return 0
+		}),
+		circuitBreakerClosedStateGauge: promauto.With(r).NewGaugeFunc(prometheus.GaugeOpts{
+			Name:        circuitBreakerCurrentStateGaugeName,
+			Help:        circuitBreakerCurrentStateGaugeHelp,
+			ConstLabels: map[string]string{circuitBreakerCurrentStateGaugeLabel: circuitbreaker.ClosedState.String()},
+		}, func() float64 {
+			if currentStateFn() == circuitbreaker.ClosedState {
+				return 1
+			}
+			return 0
+		}),
+		circuitBreakerTransitions: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
```

Comment:
We should initialize all possible states for this metric. Same for possible `result` values in the next metric.